api/v1alpha4/virtualmachine_affinity_types.go

```diff
@@ -0,0 +1,232 @@
+	Weight int32 `json:"weight"`
+
+	// ZoneSelectorTerm describes the affinity term.
+	ZoneSelectorTerm ZoneSelectorTerm `json:"zoneSelector"`
+}
+
+// VirtualMachineAffinityZoneAffinitySpec defines the affinity scheduling rules
+// related to zones.
+type VirtualMachineAffinityZoneAffinitySpec struct {
+	// +optional
+	// +listType=atomic
+
+	// RequiredDuringSchedulingIgnoredDuringExecution describes affinity
+	// requirements that must be met or the VM will not be scheduled.
+	//
+	// When there are multiple elements, the lists of zones corresponding to
+	// each term are intersected, i.e. all terms must be satisfied.
+	RequiredDuringSchedulingIgnoredDuringExecution []ZoneSelectorTerm `json:"requiredDuringSchedulingIgnoredDuringExecution,omitempty"`
+
+	// +optional
+	// +listType=atomic
+
+	// PreferredDuringSchedulingIgnoredDuringExecution describes affinity
+	// requirements that should be met, but the VM can still be scheduled if
+	// the requirement cannot be satisfied. The scheduler will prefer to schedule VMs to zones
+	// that satisfy the anti-affinity expressions specified by this field, but it may choose a zone that
+	// violates one or more of the expressions. The zone that is most preferred is the one with the
+	// greatest sum of weights.
+	//
+	// When there are multiple elements, the lists of zones corresponding to
+	// each term are intersected, i.e. all terms must be satisfied.
+	PreferredDuringSchedulingIgnoredDuringExecution []WeightedZoneSelectorTerm `json:"preferredDuringSchedulingIgnoredDuringExecution,omitempty"`
+}
+
+// VirtualMachineAntiAffinityZoneAffinitySpec defines the anti-affinity scheduling rules
+// related to zones.
+type VirtualMachineAntiAffinityZoneAffinitySpec struct {
+	// +optional
+	// +listType=atomic
+
+	// RequiredDuringSchedulingIgnoredDuringExecution describes affinity
+	// requirements that must be met or the VM will not be scheduled.
+	//
+	// When there are multiple elements, the lists of zones corresponding to
+	// each term are intersected, i.e. all terms must be satisfied.
+	RequiredDuringSchedulingIgnoredDuringExecution []ZoneSelectorTerm `json:"requiredDuringSchedulingIgnoredDuringExecution,omitempty"`
+
+	// +optional
+	// +listType=atomic
+
+	// PreferredDuringSchedulingIgnoredDuringExecution describes affinity
+	// requirements that should be met, but the VM can still be scheduled if
+	// the requirement cannot be satisfied. The scheduler will prefer to schedule VMs to zones
+	// that satisfy the anti-affinity expressions specified by this field, but it may choose a zone that
+	// violates one or more of the expressions. The node that is most preferred is the one with the
+	// greatest sum of weights.
+	//
+	// When there are multiple elements, the lists of zones corresponding to
+	// each term are intersected, i.e. all terms must be satisfied.
+	PreferredDuringSchedulingIgnoredDuringExecution []WeightedZoneSelectorTerm `json:"preferredDuringSchedulingIgnoredDuringExecution,omitempty"`
```

Comment:
@akutz , I think we need a `preferredDuringSchedulingPreferredDuringExecution` for the VM-VM affinity / anti-affinity with respect to nodes since we plan to enforce that during runtime as well?

Upstream doesn't have those [types](https://github.com/kubernetes/kubernetes/blob/148761ce2c309d52722a7125dffda1d5890e9f04/pkg/apis/core/types.go#L3237-L3268) for pod affinity, but I think we will need that specially for the VKS node VM AAF.

Reply:
@aruneshpa Did you see my comments up above?

> The initial PR should include both VM AF/AAF and Zone AF/AAF. We will omit Node AF/AAF for now as it is unclear if that is something we need/want to support. We are still awaiting a conclusion from PM discussions.

Yes, we _may_ need it, but we do not yet know.